MAINT: bump minimum NumPy version

tylerjereddy opened the pull request:
* Increase the minimum NumPy version supported by SciPy from version `1.25.2` to version `1.26.4`. This was suggested at https://github.com/scipy/scipy/pull/23006#discussion_r2094202867 but also more officially at
https://scientific-python.org/specs/spec-0000/#2025---quarter-2.

* The toolchain docs have not been updated yet in this patch, but most other stuff is covered I think? I actually don't see many test shims we can clean up as a result of this bump based on some `ripgrep` checking, but maybe they're just hiding (I think we just have less shims with newer NumPy versions really though).

I did notice a few places where our array API submodule(s) may be able to remove shims (but they may also serve `scikit-learn` these days, which is using a kind of LTS version of the SPEC IIRC).

tylerjereddy committed:
MAINT: bump min NumPy version

* Increase the minimum NumPy version supported by
SciPy from version `1.25.2` to version `1.26.4`. This
was suggested at https://github.com/scipy/scipy/pull/23006#discussion_r2094202867
but also more officially at
https://scientific-python.org/specs/spec-0000/#2025---quarter-2.

* The toolchain docs have not been updated yet in this patch,
but most other stuff is covered I think? I actually don't see
many test shims we can clean up as a result of this bump based
on some `ripgrep` checking, but maybe they're just hiding (I think
we just have less shims with newer NumPy versions really though).

tylerjereddy commented:
One of the failing CI jobs looks unrelated, but the other (32-bit) has failures that look "real," and the 32-bit job is a safe bet for where problems might show up when bumping NumPy version. One of the fails is a tolerance issue, the other looks more annoying (`nan` vs. `0`).

rgommers reviewed (commented):
Thanks Tyler, changes LGTM. A small test bump to 2.5e-14 seems fine for this:
```
E           Not equal to tolerance rtol=0, atol=1.5e-14
E           
E           Mismatched elements: 1 / 9 (11.1%)
E           Max absolute difference: 2.13162821e-14
```

rgommers committed:
TST: small atol bump in a `signal.cheby1` test, was failing on 32-bit

This was uncovered when changing the minimum numpy version from
1.25.2 to 1.26.4 in CI. Changed to the newer `xp_assert_close`,
because the legacy test function only allowed bumps by a factor
of 10 (`decimal` keyword).

rgommers committed:
DOC: add SciPy 1.16 entry in toolchain roadmap with numpy/python versions

lucascolley reviewed (commented):
not sure if the 32-bit lobpcg failure is related or just flaky

> but they may also serve scikit-learn these days, which is using a kind of LTS version of the SPEC IIRC

indeed

lucascolley commented:
looks like it failed again

rgommers commented:
Saving the full output for the `test_hermitian` failure in `lobpcg`:

```
E           x and y nan location mismatch:
E            x: array(nan, dtype=float32)
E            y: array(0)
```

<details>

```
 ________________________________ test_hermitian ________________________________

    @pytest.mark.filterwarnings("ignore:The problem size")
    def test_hermitian():
        """Check complex-value Hermitian cases.
        """
        rnd = np.random.RandomState(0)
    
        sizes = [3, 12]
        ks = [1, 2]
        gens = [True, False]
    
        for s, k, gen, dh, dx, db in (
            itertools.product(sizes, ks, gens, gens, gens, gens)
        ):
            H = rnd.random((s, s)) + 1.j * rnd.random((s, s))
            H = 10 * np.eye(s) + H + H.T.conj()
            H = H.astype(np.complex128) if dh else H.astype(np.complex64)
    
            X = rnd.standard_normal((s, k))
            X = X + 1.j * rnd.standard_normal((s, k))
            X = X.astype(np.complex128) if dx else X.astype(np.complex64)
    
            if not gen:
                B = np.eye(s)
                w, v = lobpcg(H, X, maxiter=99, verbosityLevel=0)
                # Also test mixing complex H with real B.
                wb, _ = lobpcg(H, X, B, maxiter=99, verbosityLevel=0)
                assert_allclose(w, wb, rtol=1e-6)
                w0, _ = eigh(H)
             else:
                B = rnd.random((s, s)) + 1.j * rnd.random((s, s))
                B = 10 * np.eye(s) + B.dot(B.T.conj())
                B = B.astype(np.complex128) if db else B.astype(np.complex64)
                w, v = lobpcg(H, X, B, maxiter=99, verbosityLevel=0)
                w0, _ = eigh(H, B)
    
            for wx, vx in zip(w, v.T):
                # Check eigenvector
>               assert_allclose(np.linalg.norm(H.dot(vx) - B.dot(vx) * wx)
                                / np.linalg.norm(H.dot(vx)),
                                0, atol=5e-2, rtol=0)

B          = array([[15.069941 +0.j        ,  4.268107 -1.8348355j ,
         3.0245976-0.7466751j ,  4.3077836-0.9482665j ,
      ...992333 +0.38796195j,  5.2607656+1.0951304j ,
         6.31155  -0.22033027j, 17.672451 +0.j        ]], dtype=complex64)
H          = array([[11.219306  +0.j        ,  1.2454487 -0.72869945j,
         0.76218045-0.31819183j,  1.3391619 -0.44561467j,
  ....3020495j ,  1.2695774 -0.3635519j ,
         0.9654568 +0.59222305j, 11.91809   +0.j        ]],
      dtype=complex64)
X          = array([[ 0.01751354-0.09793871j],
       [-0.00638181+0.15755548j],
       [-0.02951008-0.0948719j ],
       [ 0.04698...       [ 0.06298218+0.07232407j],
       [-0.03452053+0.0038308j ],
       [ 0.07159111-0.06361656j]], dtype=complex64)
_          = array([[-0.01746292+0.01052746j, -0.06110312+0.00032547j,
         0.05916321+0.01732739j, -0.08923098+0.01556389j,
  ....02872991j,  0.03922241+0.06509615j,
        -0.02870508-0.13998914j, -0.03576367-0.05511664j]],
      dtype=complex64)
db         = False
dh         = False
dx         = False
gen        = True
gens       = [True, False]
j          = 11
k          = 1
ks         = [1, 2]
rnd        = RandomState(MT19937) at 0xD0ED4F28
s          = 12
sizes      = [3, 12]
v          = array([[        nan+0.05720073j],
       [ 0.0075608 -0.04759835j],
       [ 0.07710112+0.06185677j],
       [-0.02873...       [-0.01180947-0.08892325j],
       [ 0.01195059+0.15658964j],
       [-0.06547873-0.00604624j]], dtype=complex64)
vx         = array([        nan+0.05720073j,  0.0075608 -0.04759835j,
        0.07710112+0.06185677j, -0.02873235-0.01361567j,
    ...965447+0.01522055j, -0.01180947-0.08892325j,
        0.01195059+0.15658964j, -0.06547873-0.00604624j], dtype=complex64)
w          = array([1.1926982], dtype=float32)
w0         = array([0.26266608, 0.5981605 , 0.6344268 , 0.68647546, 0.71724135,
       0.81510466, 0.84822   , 0.88524073, 1.0026487 , 1.0504655 ,
       1.1388186 , 1.1926988 ], dtype=float32)
wb         = array([11.93121968, 10.56323195])
wx         = 1.1926982

scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py:389: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<function assert_allclose.<locals>.compare at 0xc8feb528>, array(nan, dtype=float32), array(0))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=0, atol=0.05', 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=0, atol=0.05
E           
E           x and y nan location mismatch:
E            x: array(nan, dtype=float32)
E            y: array(0)

args       = (<function assert_allclose.<locals>.compare at 0xc8feb528>, array(nan, dtype=float32), array(0))
func       = <function assert_array_compare at 0xe615abb8>
kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=0, atol=0.05', 'verbose': True}
self       = <contextlib._GeneratorContextManager object at 0xe615c530>

/opt/python/cp311-cp311/lib/python3.11/contextlib.py:81: AssertionError
=========================== short test summary info ============================
FAILED scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py::test_hermitian
= 1 failed, 63246 passed, 3713 skipped, 16216 deselected, 755 xfailed, 12 xpassed in 784.48s (0:13:04) =
```

</details>

The `lobpcg` tests are very fragile, and it's unrelated to this PR - this test simply wasn't run in CI on 32-bit Linux with this version of numpy before. I'll open a separate issue for this and skip the test for numpy <2.0 on 32-bit systems.

rgommers committed:
TST: sparse.linalg: skip failing lobpcg test

lucascolley reviewed (approved):
CI failure looks like just a blip. Thanks all!

rgommers commented:
Yep, all happy now, useful to get this in.
